File: sdk-python/gatedhouse/audit/logger.py
"""Audit logger — publishes authorization decision events."""

from __future__ import annotations

import logging
from datetime import datetime, timezone

from gatedhouse.core.config import ResolvedConfig
from gatedhouse.core.types import AuditEntry, GatehouseEvent
from gatedhouse.events.adapters import EventBusAdapter

logger = logging.getLogger("gatedhouse.audit")


class AuditLogger:
    """Publishes audit events for authorization decisions."""

    def __init__(self, config: ResolvedConfig, event_bus: EventBusAdapter) -> None:
        self._config = config
        self._event_bus = event_bus

    async def log(self, entry: AuditEntry) -> None:
        """Log an authorization decision."""
        if not self._config.audit.enabled:
            return

        if entry.result == "denied" and not self._config.audit.log_denied:
            return
        if entry.result == "allowed" and not self._config.audit.log_allowed:
            return

        event = GatehouseEvent(
            type="gatedhouse.permission.checked",
            timestamp=datetime.now(timezone.utc).isoformat(),
            data={
                "action": entry.action,
                "result": entry.result,
                "identity_id": entry.ctx.identity.id,
                "identity_type": entry.ctx.identity.type,
                "org_id": entry.ctx.org.id,
                "membership_id": entry.ctx.membership.id,
                "resource_type": entry.resource_type,
                "resource_id": entry.resource_id,
                "reason": entry.reason,
            },
        )

        try:
            await self._event_bus.publish("gatedhouse.audit", event)
        except Exception:
            logger.exception("Failed to publish audit event")

File: sdk-typescript/src/audit/logger.ts
/**
 * Audit logger — records authorization decisions for compliance.
 *
 * Publishes audit events that can be consumed by the Audit & Compliance service.
 */

import { AuditEntry, EventBusAdapter, GatehouseEvent } from '../types';
import { ResolvedConfig } from '../config';
import { createLogger } from '../logger';

const logger = createLogger('audit');

export class AuditLogger {
  constructor(
    private config: ResolvedConfig,
    private eventBus?: EventBusAdapter,
  ) {}

  /**
   * Log an authorization decision.
   */
  async log(entry: AuditEntry): Promise<void> {
    if (!this.config.audit.enabled) return;

    // Skip allowed decisions unless logAllowed is true
    if (entry.result === 'allowed' && !this.config.audit.logAllowed) return;

    // Skip denied decisions unless logDenied is true
    if (entry.result === 'denied' && !this.config.audit.logDenied) return;

    const auditEvent: GatehouseEvent = {
      type: `authz.decision.${entry.result}`,
      timestamp: new Date().toISOString(),
      data: {
        service: this.config.service,
        action: entry.action,
        identity_type: entry.ctx.identity.type,
        identity_id: entry.ctx.identity.id,
        membership_id: entry.ctx.membership.id,
        org_id: entry.ctx.org.id,
        resource: entry.resource,
        roles: entry.ctx.roles,
        reason: entry.reason,
        ...(entry.ctx.delegation
          ? {
              delegation: {
                delegation_id: entry.ctx.delegation.id,
                delegator_id: entry.ctx.delegation.delegatorId,
              },
            }
          : {}),
      },
    };

    // Log locally
    if (entry.result === 'denied') {
      logger.warn(
        {
          action: entry.action,
          identityId: entry.ctx.identity.id,
          orgId: entry.ctx.org.id,
          reason: entry.reason,
        },
        'Access denied',
      );
    } else {
      logger.debug(
        {
          action: entry.action,
          identityId: entry.ctx.identity.id,
          orgId: entry.ctx.org.id,
        },
        'Access allowed',
      );
    }

    // Publish to event bus
    if (this.eventBus) {
      try {
        await this.eventBus.publish(auditEvent.type, auditEvent);
      } catch (err) {
        logger.error({ err }, 'Failed to publish audit event');
      }
    }
  }
}
